qiita_db/processing_job.py

```diff
@@ -6,21 +6,295 @@
+    # before it can run.
+    #
+    # Note that the main Qiita script instantiates an object of this class in
+    # a separate thread, so it can periodically update the database w/metadata
+    # from Watcher's queue. Qiita's script also calls qdb.complete() so there
+    # are no circular references. TODO: replace w/a REST call.
+
+    # valid Qiita states:
+    #             The current status of the job, one of {'queued', 'running',
+    #             'success', 'error', 'in_construction', 'waiting'}
+
+    # TODO: what to map in_construction to?
+    job_state_map = {'C': 'completed', 'E': 'exiting', 'H': 'held',
+                     'Q': 'queued', 'R': 'running', 'T': 'moving',
+                     'W': 'waiting', 'S': 'suspended'}
+
+    # TODO: moving, waiting, and suspended have been mapped to
+    # 'running' in Qiita, as 'waiting' in Qiita connotes that the
+    # main job itself has completed, and is waiting on validator
+    # jobs to finish, etc. Revisit
+    torque_to_qiita_state_map = {'completed': 'completed',
+                                 'held': 'queued',
+                                 'queued': 'queued',
+                                 'exiting': 'running',
+                                 'running': 'running',
+                                 'moving': 'running',
+                                 'waiting': 'running',
+                                 'suspended': 'running',
+                                 'DROPPED': 'error'}
+
+    def __init__(self):
+        super(Watcher, self).__init__()
+
+        # set self.owner to qiita, or whomever owns processes we need to watch.
+        self.owner = qiita_config.trq_owner
+
+        # Torque is set to drop jobs from its queue 60 seconds after
+        # completion, by default. Setting a polling value less than
+        # that allows for multiple chances to catch the exit status
+        # before it disappears.
+        self.polling_value = qiita_config.trq_poll_val
+
+        # the cross-process method by which to communicate across
+        # process boundaries. Note that when Watcher object runs,
+        # another process will get created, and receive a copy of
+        # the Watcher object. At this point, these self.* variables
+        # become local to each process. Hence, the main process
+        # can't see self.processes for example; theirs will just
+        # be empty.
+        self.queue = Queue()
+        self.processes = {}
+
+        # the cross-process sentinel value to shutdown Watcher
+        self.event = Event()
+
+    def _element_extract(self, snippet, list_of_elements,
+                         list_of_optional_elements):
+        results = {}
+        for element in list_of_elements:
+            value = search('<%s>(.*?)</%s>' % (element, element), snippet)
+            if value:
+                results[element] = value.group(1)
+            else:
+                raise AssertionError("%s element is missing!" % element)
```

Comment:
I could see the value of returning all of them. Since this code is now in a separate method, and the inputs are parameterized, it's no longer appropriate to have it be an AssertionError to begin with. I will put it as a TODO to create a Watcher class error.